src/validation.cpp

```diff
@@ -1060,28 +1176,93 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
 
     Workspace ws(ptx);
 
-    if (!PreChecks(args, ws)) return MempoolAcceptResult(ws.m_state);
+    if (!PreChecks(args, ws)) return MempoolAcceptResult(*ws.m_ptx, ws.m_state);
```

Comment:
It's not immediately obvious what the different constructor calls here indicate. Perhaps adding factory methods to the struct would make it a bit clearer:

```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index 121181393d..c65964a976 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -1176,7 +1176,7 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
 
     Workspace ws(ptx);
 
-    if (!PreChecks(args, ws)) return MempoolAcceptResult(*ws.m_ptx, ws.m_state);
+    if (!PreChecks(args, ws)) return MempoolAcceptResult::Failure(*ws.m_ptx, ws.m_state);
 
     // Only compute the precomputed transaction data if we need to verify
     // scripts (ie, other policy checks pass). We perform the inexpensive
@@ -1184,20 +1184,20 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
     // checks pass, to mitigate CPU exhaustion denial-of-service attacks.
     PrecomputedTransactionData txdata;
 
-    if (!PolicyScriptChecks(args, ws, txdata)) return MempoolAcceptResult(*ws.m_ptx, ws.m_state);
+    if (!PolicyScriptChecks(args, ws, txdata)) return MempoolAcceptResult::Failure(*ws.m_ptx, ws.m_state);
 
-    if (!ConsensusScriptChecks(args, ws, txdata)) return MempoolAcceptResult(*ws.m_ptx, ws.m_state);
+    if (!ConsensusScriptChecks(args, ws, txdata)) return MempoolAcceptResult::Failure(*ws.m_ptx, ws.m_state);
 
     // Tx was accepted, but not added
     if (args.m_test_accept) {
-        return MempoolAcceptResult(*ws.m_ptx, std::move(ws.m_replaced_transactions), ws.m_base_fees);
+        return MempoolAcceptResult::Success(*ws.m_ptx, std::move(ws.m_replaced_transactions), ws.m_base_fees);
     }
 
-    if (!Finalize(args, ws)) return MempoolAcceptResult(*ws.m_ptx, ws.m_state);
+    if (!Finalize(args, ws)) return MempoolAcceptResult::Failure(*ws.m_ptx, ws.m_state);
 
     GetMainSignals().TransactionAddedToMempool(ptx, m_pool.GetAndIncrementSequence());
 
-    return MempoolAcceptResult(*ws.m_ptx, std::move(ws.m_replaced_transactions), ws.m_base_fees);
+    return MempoolAcceptResult::Success(*ws.m_ptx, std::move(ws.m_replaced_transactions), ws.m_base_fees);
 }
 
 std::vector<MempoolAcceptResult> MemPoolAccept::AcceptMultipleTransactions(std::vector<CTransactionRef>& txns, ATMPArgs& args)
@@ -1219,8 +1219,8 @@ std::vector<MempoolAcceptResult> MemPoolAccept::AcceptMultipleTransactions(std::
             // Exit early to avoid doing pointless work. Return results in the same order as input txns.
             const auto failed_or_unfinished = [&, failed_ptx = ws.m_ptx](Workspace& ws) {
                                               return ws.m_ptx == failed_ptx
-                                              ? MempoolAcceptResult(*ws.m_ptx, ws.m_state, /* finished */ true)
-                                              : MempoolAcceptResult(*ws.m_ptx, ws.m_state, /* finished */ false);
+                                              ? MempoolAcceptResult::Failure(*ws.m_ptx, ws.m_state, /* finished */ true)
+                                              : MempoolAcceptResult::Failure(*ws.m_ptx, ws.m_state, /* finished */ false);
             };
             std::transform(workspaces.begin(), workspaces.end(), std::back_inserter(results), failed_or_unfinished);
             return results;
@@ -1240,15 +1240,15 @@ std::vector<MempoolAcceptResult> MemPoolAccept::AcceptMultipleTransactions(std::
             CTransactionRef failed_ptx = ws.m_ptx;
             const auto failed_or_unfinished = [&failed_ptx](Workspace& ws) {
                                               return ws.m_ptx == failed_ptx
-                                              ? MempoolAcceptResult(*ws.m_ptx, ws.m_state, /* finished */ true)
-                                              : MempoolAcceptResult(*ws.m_ptx, ws.m_state, /* finished */ false);
+                                              ? MempoolAcceptResult::Failure(*ws.m_ptx, ws.m_state, /* finished */ true)
+                                              : MempoolAcceptResult::Failure(*ws.m_ptx, ws.m_state, /* finished */ false);
             };
             auto it_curr = std::find_if(workspaces.begin(), workspaces.end(),
                                         [& failed_ptx](Workspace& ws) { return ws.m_ptx == failed_ptx; });
             // When test_accept=true, transactions that pass PolicyScriptChecks are valid because there are
             // no further mempool checks (passing PolicyScriptChecks implies passing ConsensusScriptChecks).
             std::transform(workspaces.begin(), it_curr, std::back_inserter(results), [](Workspace& ws) {
-                           return MempoolAcceptResult(*ws.m_ptx, std::move(ws.m_replaced_transactions), ws.m_base_fees);
+                           return MempoolAcceptResult::Success(*ws.m_ptx, std::move(ws.m_replaced_transactions), ws.m_base_fees);
             });
             std::transform(it_curr, workspaces.end(), std::back_inserter(results), failed_or_unfinished);
             return results;
@@ -1260,7 +1260,7 @@ std::vector<MempoolAcceptResult> MemPoolAccept::AcceptMultipleTransactions(std::
 
     std::transform(workspaces.begin(), workspaces.end(), std::back_inserter(results), [](Workspace& ws) {
         // All successful MemPoolAcceptResults
-        return MempoolAcceptResult(*ws.m_ptx, std::move(ws.m_replaced_transactions), ws.m_base_fees);
+        return MempoolAcceptResult::Success(*ws.m_ptx, std::move(ws.m_replaced_transactions), ws.m_base_fees);
     });
     return results;
 }
diff --git a/src/validation.h b/src/validation.h
index 64f1b24743..234c1db895 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -209,8 +209,19 @@ struct MempoolAcceptResult {
     /** Raw base fees in satoshis. */
     const std::optional<CAmount> m_base_fees;
 
+    static MempoolAcceptResult Success(const CTransaction& tx, std::list<CTransactionRef>&& replaced_txns, CAmount fees)
+    {
+        return MempoolAcceptResult(tx, std::move(replaced_txns), fees);
+    }
+
+    static MempoolAcceptResult Failure(const CTransaction& tx, TxValidationState state, bool finished=true)
+    {
+        return MempoolAcceptResult(tx, state, finished);
+    }
+
+private:
     /** Constructor for failure case */
-    explicit MempoolAcceptResult(const CTransaction& tx, TxValidationState state, bool finished=true)
+    explicit MempoolAcceptResult(const CTransaction& tx, TxValidationState state, bool finished)
         : m_tx(tx), m_result_type(finished ? ResultType::INVALID : ResultType::UNFINISHED),
         m_state(state), m_replaced_transactions(nullopt), m_base_fees(nullopt) {
             if (finished) Assume(!state.IsValid()); // Can be invalid or error
```

Reply:
Ooh that's nice! 👀